sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -42,17 +42,81 @@ import org.apache.spark.sql.types.DataType
  * @since 1.3.0
  */
 @InterfaceStability.Stable
-case class UserDefinedFunction protected[sql] (
-    f: AnyRef,
-    dataType: DataType,
-    inputTypes: Option[Seq[DataType]]) {
+class UserDefinedFunction protected[sql] (
+    val f: AnyRef,
+    val dataType: DataType,
+    val inputTypes: Option[Seq[DataType]],
+    val name: Option[String] = None,
+    val deterministic: Boolean = true,
+    val distinctLike: Boolean = false) {
```

Comment:
Since this breaks bin. compatibility (I know you are familiar with the point), the target of this pr is `spark-v3.0.0`?